runtime/contract_update_validation.go

```diff
@@ -0,0 +1,456 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package runtime
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+)
+
+type ContractUpdateValidator struct {
+	oldProgram     *ast.Program
+	newProgram     *ast.Program
+	newDeclaration ast.Declaration
+	currentDecl    ast.Declaration
+	visited        map[ast.Declaration]bool
+}
+
+// ContractUpdateValidator should implement ast.TypeEqualityChecker
+var _ ast.TypeEqualityChecker = &ContractUpdateValidator{}
+
+func NewContractUpdateValidator(oldProgram *ast.Program, newProgram *ast.Program) *ContractUpdateValidator {
+	return &ContractUpdateValidator{
+		oldProgram: oldProgram,
+		newProgram: newProgram,
+		visited:    map[ast.Declaration]bool{},
+	}
+}
+
+func (validator *ContractUpdateValidator) Validate() error {
+	oldRootDecl, err := getRootDeclaration(validator.oldProgram)
+	if err != nil {
+		return err
+	}
+
+	newRootDecl, err := getRootDeclaration(validator.newProgram)
+	if err != nil {
+		return err
+	}
+
+	validator.newDeclaration = newRootDecl
+
+	return validator.checkDeclarationUpdatability(oldRootDecl, newRootDecl)
+}
+
+func (validator *ContractUpdateValidator) checkDeclarationUpdatability(
+	oldDeclaration ast.Declaration,
+	newDeclaration ast.Declaration,
+) error {
+
+	// Do not allow converting between different types of composite declarations:
+	// e.g: - 'contracts' and 'contract-interfaces',
+	//      - 'structs' and 'enums'
+	if oldDeclaration.DeclarationKind() != newDeclaration.DeclarationKind() {
+		return &InvalidDeclarationKindChangeError{
+			name:    oldDeclaration.DeclarationIdentifier().Identifier,
+			oldKind: oldDeclaration.DeclarationKind().Name(),
+			newKind: newDeclaration.DeclarationKind().Name(),
+			Range:   ast.NewRangeFromPositioned(newDeclaration),
+		}
+	}
+
+	parentDecl := validator.currentDecl
+	validator.currentDecl = newDeclaration
+	defer func() {
+		validator.currentDecl = parentDecl
+	}()
+
+	// If the same decl is already visited, then do not check again.
+	// This also avoids getting stuck on circular dependencies between composite decls.
+	if validator.visited[newDeclaration] {
+		return nil
+	}
+	validator.visited[newDeclaration] = true
+
+	err := validator.checkFields(oldDeclaration, newDeclaration)
+	if err != nil {
+		return err
+	}
+
+	err = validator.checkNestedDeclarations(oldDeclaration, newDeclaration)
+	if err != nil {
+		return err
+	}
+
+	switch newDecl := newDeclaration.(type) {
+	case *ast.CompositeDeclaration:
+		oldDecl := oldDeclaration.(*ast.CompositeDeclaration)
+		return validator.checkConformances(oldDecl, newDecl)
+	}
```

Comment:
The type-switch with the single case can be shortened to an if-statement.
Also, you maybe want to also do the conditional cast for the old declaration.
Finally, maybe only return if an error occurs instead of always:
There might be code added in the future after this check, which would be skipped 

```suggestion
	if  newDecl, ok := newDeclaration.(*ast.CompositeDeclaration); ok {
		if oldDecl, ok := oldDeclaration.(*ast.CompositeDeclaration); ok {
			err := validator.checkConformances(oldDecl, newDecl)
			if err != nil {
				return err
			}
		}
	}
```